File: Katas/Kata02-KarateChop/java/src/search/IterativeChop.java
package search;

public class IterativeChop implements Chop {
    public int chop(int target, int[] array) {
        int top, bottom, half;

        bottom = 0;
        top = array.length - 1;
        
        if (array[bottom] == target) {
            return bottom;
        } else if (array[top] == target) {
            return top;
        }
        
        while (bottom <= top) {
            half = (top + bottom) / 2;
            if (array[half] < target) {
                bottom = half + 1;
            } else if (array[half] > target) {
                top = half - 1;
            } else {
                return half;
            }
        }

        return -1;
    }
}

File: Katas/Kata02-KarateChop/c++/src/iterative_chop.cc
/**
 * @file iterative_chop.cc
 *
 * @brief
 *  
 * @author: Jordi Sánchez, jorsanpe@gmail.com
 */
#include <recursive_chop.h>

int RecursiveChop::rechop(int value, int *array, int low, int top)
{
    int middle;

    if (low > top) {
        return -1;
    }
    middle = (top + low) / 2;
    if (value < array[middle]) {
        return this->rechop(value, array, low, middle-1);
    }
    if (value > array[middle]) {
        return this->rechop(value, array, middle+1, top);
    }
    return middle;
}

int RecursiveChop::chop(int value, int *array, int size)
{
    return this->rechop(value, array, 0, size-1);
}
